source/common/http/http1/conn_pool.cc

```diff
@@ -209,25 +215,48 @@ void ConnPoolImpl::onResponseComplete(ActiveClient& client) {
     host_->cluster().stats().upstream_cx_max_requests_.inc();
     onDownstreamReset(client);
   } else {
-    processIdleClient(client);
+    // Upstream connection might be closed right after response is complete. Setting delay=true
+    // here to attach pending requests in next dispatcher loop to handle that case.
+    // https://github.com/envoyproxy/envoy/issues/2715
+    processIdleClient(client, true);
+  }
+}
+
+void ConnPoolImpl::onUpstreamReady() {
+  upstream_ready_enabled_ = false;
+  while (!pending_requests_.empty() && !ready_clients_.empty()) {
+    ActiveClient& client = *ready_clients_.front();
+    ENVOY_CONN_LOG(debug, "attaching to next request", *client.codec_client_);
+    // There is work to do so bind a request to the client and move it to the busy list. Pending
+    // requests are pushed onto the front, so pull from the back.
+    attachRequestToClient(client, pending_requests_.back()->decoder_,
+                          pending_requests_.back()->callbacks_);
+    pending_requests_.pop_back();
+    client.moveBetweenLists(ready_clients_, busy_clients_);
   }
 }
 
-void ConnPoolImpl::processIdleClient(ActiveClient& client) {
+void ConnPoolImpl::processIdleClient(ActiveClient& client, bool delay) {
   client.stream_wrapper_.reset();
-  if (pending_requests_.empty()) {
-    // There is nothing to service so just move the connection into the ready list.
+  if (pending_requests_.empty() || delay) {
+    // There is nothing to service or delayed processing is requested, so just move the connection
+    // into the ready list.
     ENVOY_CONN_LOG(debug, "moving to ready", *client.codec_client_);
     client.moveBetweenLists(busy_clients_, ready_clients_);
```

Comment:
@lizan Too early to mark the client ready in this branch. Expecting anther read attempt to see if there is `EOF`. This client could be immediately used by another down stream request.
At the next cycle, dispatcher always handle write attempt first, even there is EOF in the read buffer, that is a 503 for this upstream request.

Reply:
At the end of the upstream read event when we reach EOF the client will be removed so this is safe. https://github.com/envoyproxy/envoy/blob/v1.10.0/source/common/network/connection_impl.cc#L486